test/videoframeprocessor/DefaultVideoTransformDevice.test.ts

```diff
@@ -0,0 +1,417 @@
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      device.addObserver(obs);
+      await device.transformStream(mockVideoStream);
+      await startCallback;
+      domMockBehavior.createElementCaptureStream = mockVideoStream;
+      assert.exists(device.outputMediaStream);
+      await device.stop();
+
+      await stopCallback;
+    });
+
+    it('can stop device idempotently', async () => {
+      const obs = new MockObserver();
+      const startCallback = called(obs.processingDidStart);
+      const stopCallback = called(obs.processingDidStop);
+
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      device.addObserver(obs);
+      domMockBehavior.createElementCaptureStream = mockVideoStream;
+      await device.transformStream(mockVideoStream);
+
+      await startCallback;
+      assert.exists(device.outputMediaStream);
+
+      await device.stop();
+
+      await stopCallback;
+      expect(device.outputMediaStream).to.be.null;
+      await device.stop();
+      expect(device.outputMediaStream).to.be.null;
+    });
+  });
+
+  describe('getInnerDevice', () => {
+    it('can get the inner device', async () => {
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      expect(device.getInnerDevice()).to.equal('test');
+    });
+  });
+
+  describe('intrinsicDevice', () => {
+    it('returns the intrinsicDevice', async () => {
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      expect(JSON.stringify(await device.intrinsicDevice())).to.equal(
+        JSON.stringify({ deviceId: { exact: 'test' } })
+      );
+    });
+
+    it('returns the correct intrinsicDevice', async () => {
+      domMockBehavior = new DOMMockBehavior();
+      domMockBehavior.isUnifiedPlanSupported = false;
+      domMockBehavior.browserName = 'ios12.1';
+      domMockBuilder = new DOMMockBuilder(domMockBehavior);
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      expect(JSON.stringify(await device.intrinsicDevice())).to.equal(
+        JSON.stringify({ deviceId: 'test' })
+      );
+    });
+
+    it('returns the correct intrinsicDevice', async () => {
+      const device = new DefaultVideoTransformDevice(logger, null, []);
+      expect(JSON.stringify(await device.intrinsicDevice())).to.equal(JSON.stringify({}));
+    });
+
+    it('returns the correct intrinsicDevice', async () => {
+      const device = new DefaultVideoTransformDevice(logger, mockVideoStream, []);
+      expect(JSON.stringify(await device.intrinsicDevice())).to.equal(
+        JSON.stringify(mockVideoStream)
+      );
```

Comment:
Is there a reason you don't use `to.deep.equal` for object comparison? (Throughout)

Reply:
changed to `deep.equal`